### PR TITLE
add copy dagservice function

### PR DESCRIPTION
**Pull request by kjzz**
feature #40

### CODE REVIEW

File: daghelpers.go [hunk 1]
```diff
@@ -89,6 +89,26 @@ func GetNodes(ctx context.Context, ds NodeGetter, keys []*cid.Cid) []*NodePromis
...
 }
 
+func Copy(ctx context.Context, from, to DAGService, root *cid.Cid) error {
+	node, err := from.Get(ctx, root)
+	if err != nil {
+		return err
+	}
+	links := node.Links()
+	if len(links) == 0 {
+		n := node.Copy()
```
Comment (Stebalien) [inline daghelpers.go]: FYI, you don't actually need to call `Copy` on the node itself. We use that when *editing* nodes (i.e., `Copy` then edit).

File: daghelpers.go [hunk 2]
```diff
@@ -89,6 +89,26 @@ func GetNodes(ctx context.Context, ds NodeGetter, keys []*cid.Cid) []*NodePromis
...
+		to.Add(ctx, n)
+		return nil
+	}
+	for _, link := range links {
```
Comment (Stebalien) [inline daghelpers.go]: We still need to copy the intermediate nodes, right? This just copies the leaves.

File: daghelpers_test.go [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+
+import (
+	"context"
+	"github.com/ipfs/go-cid"
+	mh "github.com/multiformats/go-multihash"
+	"testing"
+)
+
+func TestCopy(t *testing.T) {
+	ctx, cancel := context.WithCancel(context.Background())
```
Comment (Stebalien) [inline daghelpers_test.go]: Should probably test with a dag that has children.

**Comment by kjzz** [comment]
Hey @Stebalien , thx for your advice.I have update my pr.please help me review it?

File: daghelpers.go [hunk 4]
```diff
@@ -89,6 +89,32 @@ func GetNodes(ctx context.Context, ds NodeGetter, keys []*cid.Cid) []*NodePromis
...
+			return err
+		}
+		return nil
+	}
+	for _, link := range links {
+		err := Copy(ctx, from, to, link.Cid)
+		if err != nil {
+			return err
+		}
+		err = to.Add(ctx, node)
```
Comment (Stebalien) [inline daghelpers.go]: We don't need to do this once *per link*, just once.

File: daghelpers_test.go [hunk 5]
```diff
@@ -0,0 +1,144 @@
+package format
+
+import (
+	"context"
+	"github.com/ipfs/go-cid"
+	mh "github.com/multiformats/go-multihash"
+	"github.com/stretchr/testify/assert"
```
Comment (Stebalien) [inline daghelpers_test.go]: We try not to add *too* many external deps unless we need them. I'd prefer it if we could stick with go's testing tools.

**Comment by kjzz** [comment]
@Stebalien hello,thank u very much.And i have update pr.drop `len(links) == 0` judge.

**Comment by kjzz** [comment]
@Stebalien  hello,thank u very much.And i have update pr.drop len(links) == 0 judge.please help me review again,thx a lot

**Comment by Stebalien** [comment]
Looks good. We can improve the performance a bit (e.g., deduplicate identical sub-graphs) but this is a good start for now.

**Pull request merged**